servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpTransportObserverTest.java

```diff
@@ -0,0 +1,409 @@
+        if (protocol == Protocol.HTTP_2) {
+            // HTTP/1.x has a single write publisher across all requests that does not complete after each response
+            verify(serverWriteObserver).writeComplete();
+        }
+
+        verify(clientReadObserver, atLeastOnce()).requestedToRead(anyLong());
+        verify(clientReadObserver, atLeastOnce()).itemRead();
+        verify(clientReadObserver).readComplete();
+
+        verifyNoMoreInteractions(
+                clientDataObserver, clientMultiplexedObserver, clientReadObserver, clientWriteObserver,
+                serverDataObserver, serverMultiplexedObserver, serverReadObserver, serverWriteObserver);
+    }
+
+    @Test
+    public void serverFailsResponsePayloadBodyBeforeRead() throws Exception {
+        testServerFailsResponsePayloadBody(SVC_ERROR_BEFORE_READ, false);
+    }
+
+    @Test
+    public void serverFailsResponsePayloadBodyDuringRead() throws Exception {
+        testServerFailsResponsePayloadBody(SVC_ERROR_DURING_READ, true);
+    }
+
+    public void testServerFailsResponsePayloadBody(String path, boolean serverReadCompletes) throws Exception {
+        processRequest.countDown();
+        StreamingHttpConnection connection = streamingHttpConnection();
+        StreamingHttpResponse response = makeRequest(connection.post(path)
+                .addHeader(CONTENT_LENGTH, ZERO));
+        assertThat(response.status(), is(OK));
+        assertThat(response.version(), is(protocol.version));
+
+        ExecutionException e = assertThrows(ExecutionException.class,
+                () -> response.payloadBody().ignoreElements().toFuture().get());
+        Class<? extends Throwable> causeType = protocol == Protocol.HTTP_1 ?
+                ClosedChannelException.class : H2StreamResetException.class;
+        assertThat(e.getCause(), instanceOf(causeType));
+
+        if (protocol == Protocol.HTTP_2) {
+            connection.closeGracefully();
+        }
+        assertConnectionClosed();
+        verifyNewReadAndNewWrite(1);
+
+        verify(clientWriteObserver, atLeastOnce()).requestedToWrite(anyLong());
+        verify(clientWriteObserver, atLeastOnce()).itemReceived();
+        verify(clientWriteObserver, atLeastOnce()).onFlushRequest();
+        verify(clientWriteObserver, atLeastOnce()).itemWritten();
+        verify(clientWriteObserver).writeComplete();
+        // Failure of the read triggers cancellation of the write.
+        verify(clientWriteObserver, await()).writeCancelled();
+
+        verify(serverReadObserver, atLeastOnce()).requestedToRead(anyLong());
+        verify(serverReadObserver, atLeastOnce()).itemRead();
+        if (serverReadCompletes) {
+            verify(serverReadObserver).readComplete();
+        } else {
+            // FIXME: because nobody subscribes to the request payload publisher, the terminal signal is not delivered
+            // verify(serverReadObserver).readFailed(any(ClosedChannelException.class));
```

Comment:
@NiteshKant @Scottmitch wdyt about this?

The problem here is that connection is closed, but the `connection.read()` did not terminate and was not canceled. This happens because service did not subscribe to the payload body publisher and auto-draining does not happen if response body publisher fails because the connection (or stream) will be closed anyway. Therefore, `NettyChannelPublisher#pending` queue has 2 elements:
1. empty trailers
2. `ClosedChannelException`

Because there are trailers, the terminal event is not delivered. IIUC the control flow the terminal signal is delayed intentionally to let the subscriber consume a successful request before terminating it.

Is it ok that we do not terminate `connection.read()` in this case when connection (or stream) is closed?

Reply:
> the connection.read() did not terminate and was not canceled. 

The server side subscribes to `connection.read()` to read the request so it should be cancelled if the response failed which in turn shuts down the connection processing. Looks like something is missing in this analysis/test.

Reply:
Agreed, `readCancel` is expected here instead, but does not happen:
```
Wanted but not invoked:
serverReadObserver.readCancelled();
-> at io.servicetalk.http.netty.HttpTransportObserverTest.testServerFailsResponsePayloadBody(HttpTransportObserverTest.java:310)

However, there were exactly 2 interactions with this mock:
serverReadObserver.requestedToRead(1L);
-> at io.servicetalk.transport.netty.internal.DefaultNettyConnection$2$1.lambda$request$0(DefaultNettyConnection.java:291)

serverReadObserver.itemRead();
-> at io.servicetalk.transport.netty.internal.TransportObserverUtils.safeReport(TransportObserverUtils.java:124)
```

Opened #1112